{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Motif Marker:\n",
    "## Adrian Bubie\n",
    "## 2/11/18\n",
    "## --------------------\n",
    "## This program detects motifs (selected sequences) of interest from sequences provided in a fasta format. \n",
    "## As the intended use case is to identify motifs around exon boundries that are potential regulators for splicing events, \n",
    "## this program maps detected motifs onto a representation of the intron/exons sequences provided in the input file, and\n",
    "## exports these maps as an SVG graphic.\n",
    "\n",
    "import textwrap\n",
    "import random\n",
    "import re\n",
    "import argparse as ap\n",
    "import cairo\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_arguments():\n",
    "    parser = ap.ArgumentParser(prog='./Motif_Marker.py', formatter_class=ap.RawDescriptionHelpFormatter, description=textwrap.dedent('''\\\n",
    "    Motif Sequence Marker\n",
    "    ---------------------\n",
    "    Locates and counts specified motif sequences from given gene sequences\n",
    "    in FASTA format, then maps motifs' locations in relation to Intron/Exon of given sequence.\n",
    "    Produces an .svg image file with each a 1-to-1 px/bp graph of sequences with motifs indicated by colored bars.\n",
    "    \n",
    "    Takes input of FASTA file with any exonic sequences capitalized, and intronic sequences lowercase.\n",
    "    Requires motifs to search for and size of intronic sequence flanking exons to be searched to be specified by the user. \n",
    "    '''))\n",
    "    parser.add_argument(\"-f\", help=\"Fasta file to be processed. Exon sequences must be capitalized, and gene names contained in sequence ID. Must include absolute path the file. <str>\", required=True, type=str)\n",
    "    parser.add_argument(\"-w\", help=\"Intronic flanking sequence window sized (applies to both sides of exonic regions). Default is 200bp. <int>\", required=False, type=int, default=200)\n",
    "    parser.add_argument(\"-m\", help=\"Motif file to define motif sequences to query for (one Motif per line). Motifs must use *IUPAC* Nucleotide codes. Must include absolute path to file <str>\", required=True, type=str)\n",
    "    parser.add_argument(\"-s\", help=\"Optional summary output file containing the count of each motif per fasta sequence. File is created in location of imput file if set to 'True' <str>\", required=False, type=bool, default=False)\n",
    "    parser.add_argument(\"-title\", help=\"Option Title for the Motif Graph. If not provided, Graph is produced without a title <str>\", required=False, type=str, default='')\n",
    "    #parser.add_argument(\"-colors\", help=\"Set color palate to use for motif markers (note: whole sequences will always be represented in black). Provide as list of hex codes (#)\", required=False, type=str, default='')\n",
    "    return parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def iupac_interp(motifs_file):\n",
    "    '''IUPAC Interpreter: Takes motif file, parses out motifs, and returns list of regex search terms for each motif\n",
    "       based on IUPAC nomencalture.'''\n",
    "    \n",
    "    # Keys of dict are nucleotide code, and values are bases:\n",
    "    iupac = {\"A\":'[Aa]',\"C\":'[Cc]',\"G\":'[Gg]',\"T\":'[TUtu]',\"U\":'[TUtu]',\"R\":'[AGag]',\"Y\":'[CTct]',\"S\":'[GCgc]',\"W\":'[ATat]',\"K\":'[GTgt]',\"M\":'[ACac]',\"B\":'[CGTcgt]',\"D\":'[AGTagt]',\"H\":'[ACTact]',\"V\":'[ACGacg]',\"N\":'[A-Za-z]',\n",
    "             \"a\":'[Aa]',\"c\":'[Cc]',\"g\":'[Gg]',\"t\":'[TUtu]',\"u\":'[TUtu]',\"r\":'[AGag]',\"y\":'[CTct]',\"s\":'[GCgc]',\"w\":'[ATat]',\"k\":'[GTgt]',\"m\":'[ACac]',\"b\":'[CGTcgt]',\"d\":'[AGTagt]',\"h\":'[ACTact]',\"v\":'[ACGacg]',\"n\":'[A-Za-z]'}\n",
    "    \n",
    "    with open(motifs_file, 'r') as mtfs:\n",
    "        search_terms = []       # Create a list to store the returned search terms\n",
    "        untrans_terms = []      # Keep a list of the untranslated strings for plotting\n",
    "        line = mtfs.readline()  \n",
    "        while line:             # For each motif in the file\n",
    "            st = ''\n",
    "            mt = str(line).strip('\\n')\n",
    "            untrans_terms.append(mt)\n",
    "            for char in mt:             # For each character in the motif\n",
    "                if char in iupac.keys():\n",
    "                    st = st+iupac[char]   # If the character is in the IUPAC dict, add it to the current search term\n",
    "                else:\n",
    "                    raise ValueError('Error: motif contains character not in IUPAC nucleotide codes; motif cannot be translated') # Throw exception if the motif contains character not in IUPAC\n",
    "            \n",
    "            search_terms.append(st)\n",
    "            line = mtfs.readline()\n",
    "    \n",
    "    return search_terms, untrans_terms       # Return the search terms\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class fasta_sequence():\n",
    "    '''Fasta Sequence object: collection of sequence lines for a single fasta ID.\n",
    "       Object collects lines into single string for motif searching, collects seq length, and more.'''\n",
    "    \n",
    "    def __init__(self, lines):\n",
    "        self.gene = lines[0].split(' ')[0].strip('>')\n",
    "        self.seq = ''.join([line.strip('\\n') for line in lines if line.startswith('>') != True])\n",
    "    \n",
    "    def tot_seq_len(self):\n",
    "        return int(len(self.seq))\n",
    "    \n",
    "    def exon_bounds(self):\n",
    "        exon_st = self.seq.find(re.search('[ATCG]+',self.seq)[0])\n",
    "        exon_ed = (self.seq[exon_st:].find(re.search('[atcg]+',self.seq[exon_st:])[0]))+ exon_st\n",
    "        return [exon_st, exon_ed]\n",
    "    \n",
    "    def exon_bounds_rel(self, trimmed_seq):\n",
    "        exon_st = trimmed_seq.find(re.search('[ATCG]+',trimmed_seq)[0])\n",
    "        exon_ed = (trimmed_seq[exon_st:].find(re.search('[atcg]+',trimmed_seq[exon_st:])[0]))+ exon_st\n",
    "        return [exon_st, exon_ed]\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def window_trim(fasta, window_size):\n",
    "    '''Window Trimmer: Takes in a fasta gene sequence and trims the flagging intronic sequences surrounding the exonic,\n",
    "       uppercase sequence to the appropriate window size. If the window size is bigger than the intronic segments, the \n",
    "       entire sequence is returned.'''\n",
    "    \n",
    "    exon_bounds = fasta.exon_bounds()\n",
    "    \n",
    "    if len(fasta.seq[:exon_bounds[0]]) > window_size:\n",
    "        search_start = len(fasta.seq[:exon_bounds[0]])-window_size\n",
    "    else:\n",
    "        search_start = 0\n",
    "    \n",
    "    if len(fasta.seq[exon_bounds[1]:]) > window_size:\n",
    "        search_end = exon_bounds[1]+window_size\n",
    "    else:\n",
    "        search_end = fasta.tot_seq_len()\n",
    "    \n",
    "    trimmed_seq = fasta.seq[search_start:search_end]\n",
    "    \n",
    "    return trimmed_seq\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def Motif_search(fasta, window, motifs):\n",
    "    '''Motif Search: Takes in a fasta gene sequence and list of motifs to be searched for, then searches the trimmed\n",
    "       version of the sequence for each motif in the set of motifs, and saves the motif positions to a dictionary. Returns\n",
    "       this dictionary of search results.'''\n",
    "    \n",
    "    # Start by getting the window trimmed seq:\n",
    "    w_seq = window_trim(fasta, window)\n",
    "    \n",
    "    # For each motif, search the sequence for the positions of these motifs, and store the positions:\n",
    "    motif_positions = {}\n",
    "    for i in range(0,len(motifs[0])):\n",
    "        motif_positions[motifs[0][i]]=[x.start(0) for x in re.finditer(motifs[0][i], w_seq)]\n",
    "        motif_positions[motifs[0][i]].append(motifs[1][i])\n",
    "\n",
    "    return motif_positions\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def py_draw(search_results, Motifs, window, title):\n",
    "    '''Pycairo Graphing: Takes in the motif search results and draws to-scale visual pertaining to the sequence, exons\n",
    "       and motifs found. Produces an .svg of the graph in the current directory (where script is executed from)'''\n",
    "    \n",
    "    no_of_graphs = len(search_results) # Number of graphs to be drawn\n",
    "    \n",
    "    max_size = 0\n",
    "    for res in search_results:\n",
    "        if len(window_trim(res[0],window)) > max_size:\n",
    "            max_size = len(window_trim(res[0],window))  # Get the sequence length you have to determine surface width\n",
    "    \n",
    "    # Get Motif color schemes for legend:\n",
    "    col = 1\n",
    "    motif_color = {}\n",
    "    for motif in Motifs[0]:\n",
    "        r = 0.2+(col/5)\n",
    "        g = 0.8-(col/5)\n",
    "        b = 0.2+(col/10)\n",
    "        motif_color[motif]=[r,g,b]\n",
    "        col = col+1 \n",
    "    \n",
    "    # Start Graph drawing\n",
    "    surface = cairo.SVGSurface(\"./exon_graphs.svg\", max_size+45, (no_of_graphs*60)+100) # width, height for dimensions\n",
    "    context = cairo.Context(surface)\n",
    "    \n",
    "    # Graph Title, if indicated:\n",
    "    if title is not '':\n",
    "        context.set_font_size(16)\n",
    "        context.move_to(5,35)\n",
    "        context.show_text(title)\n",
    "        \n",
    "    \n",
    "    for i in range(0,len(search_results)):    # For each sequence with results:\n",
    "        context.set_line_width(1)\n",
    "        \n",
    "        # Add Gene name before graph\n",
    "        context.set_font_size(10)\n",
    "        context.set_source_rgb(0, 0, 0)\n",
    "        context.select_font_face(\"Courier\", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_BOLD)\n",
    "        context.move_to(5,30*(i+1)+50)  # x,y coords; spacing graphs by 30px\n",
    "        context.show_text(search_results[i][0].gene)\n",
    "        \n",
    "        # Draw Line to represent Sequence\n",
    "        context.move_to(35,30*(i+1)+50) # x,y coords; spacing graphs by 30px\n",
    "        context.line_to(len(window_trim(search_results[i][0],window))+35,30*(i+1)+50) # Drawing line px length of sequence\n",
    "        context.stroke()\n",
    "        \n",
    "        # Draw Rectangle to represent Exon\n",
    "        exon_coords = search_results[i][0].exon_bounds_rel(window_trim(search_results[i][0],window))  # Get coordinates for the exon in the sequence\n",
    "        context.rectangle(exon_coords[0]+35,(30*(i+1)+50)-10,exon_coords[1]-exon_coords[0],18)       # x,y top corner, followed by width, height of rec.\n",
    "        context.fill()\n",
    "        \n",
    "        # Add Motifs as small rectangles in appropriate locs\n",
    "        col = 1\n",
    "        for key in search_results[i][1].keys():\n",
    "            context.set_source_rgb(motif_color[key][0],motif_color[key][1],motif_color[key][2])\n",
    "            mot_width = key.count('[')\n",
    "            for loc in search_results[i][1][key]:\n",
    "                if type(loc) == int:\n",
    "                    context.rectangle(loc+35,(30*(i+1)+50)-10,mot_width,20)\n",
    "                    context.fill()\n",
    "            col = col+1\n",
    "    \n",
    "    # Draw legend\n",
    "    context.move_to(10,(no_of_graphs*50))\n",
    "    context.set_source_rgb(0, 0, 0)\n",
    "    context.show_text(\"Key\")\n",
    "    context.select_font_face(\"Courier\", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_NORMAL)\n",
    "    for i in range(0,len(Motifs[0])):\n",
    "        context.set_source_rgb(motif_color[Motifs[0][i]][0],motif_color[Motifs[0][i]][1],motif_color[Motifs[0][i]][2])\n",
    "        context.rectangle(10,(no_of_graphs*50)+((i+1)*15),5,10)\n",
    "        context.fill()\n",
    "        \n",
    "        context.move_to(20,(no_of_graphs*50)+((i+1)*15)+7)\n",
    "        context.set_source_rgb(0, 0, 0)\n",
    "        context.show_text(search_results[0][1][Motifs[0][i]][-1])\n",
    "        context.move_to(10,(no_of_graphs*50)+((i+1)*15))\n",
    "        \n",
    "    surface.finish()\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def summary_out(results):\n",
    "    '''Summary File: Takes in the motif search results and produces a text file of the number and location of each motif\n",
    "       type for each of the searched sequences in the fasta. Summary file is written out to the current directory \n",
    "       (where script is executed from)'''\n",
    "    \n",
    "    with open('./Motif_Search_Summary.txt','w') as out:\n",
    "        out.write('## Summary of Motif Search:\\n')\n",
    "        for res in results:\n",
    "            out.write('Gene: '+res[0].gene+'\\n')\n",
    "            for key in res[1].keys():\n",
    "                out.write('\\tMotif: '+res[1][key][-1]+'\\n')\n",
    "                out.write('\\tNumber found: '+str(len(res[1][key])-1)+'\\n')\n",
    "                out.write('\\tPosition(s) in sequence: '+str(res[1][key][:-1])+'\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "####################\n",
    "## Main function: ##\n",
    "####################\n",
    "\n",
    "# Fasta = open(args.f,'r')\n",
    "# window = args.w\n",
    "# Motifs = iupac_interp(args.m)\n",
    "# title = args.title\n",
    "\n",
    "\n",
    "Motifs = iupac_interp('/Users/Adrian/BGMP/Motif_marker/test/motifs.txt')\n",
    "Fasta = open('/Users/Adrian/BGMP/Motif_marker/test/fasta_t2.fa','r')\n",
    "title =''\n",
    "line = Fasta.readline()\n",
    "seq_results = []\n",
    "\n",
    "while line:\n",
    "    to_process = []\n",
    "    to_process.append(line)\n",
    "    line = Fasta.readline()\n",
    "    while (line.startswith('>') != True) and line != '':\n",
    "        to_process.append(line)\n",
    "        line = Fasta.readline()\n",
    "    \n",
    "    results = Motif_search(fasta_sequence(to_process), 200, Motifs)\n",
    "    \n",
    "    seq_results.append([fasta_sequence(to_process),results])\n",
    "\n",
    "Fasta.close()\n",
    "\n",
    "#if args.s == True:\n",
    "summary_out(seq_results)\n",
    "    \n",
    "# Call function to draw motif graphs:\n",
    "py_draw(seq_results, Motifs, 200, title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
